api/hasura/actions/vouch.ts

```diff
@@ -0,0 +1,181 @@
+    );
+  }
+
+  const { vouches } = await gql.getExistingVouch(nomineeId, voucher.id);
+
+  if (vouches.pop()) {
+    throw new ForbiddenError('voucher has already vouched for this nominee');
+  }
+
+  return {
+    voucher,
+  };
+}
+
+async function vouch(nomineeId: number, voucher: voucher) {
+  // vouch for the nominee
+
+  // this inserts the vouch and also fetches the nominee with updated vouch count
+  const insert_vouches = await gql.insertVouch(nomineeId, voucher.id);
+  if (!insert_vouches?.nominee) {
+    throw new InternalServerError('unable to add vouch');
+  }
+
+  const nominee = insert_vouches.nominee;
+
+  // announce the vouching
+  await sendSocialMessage({
+    message: `${nominee.name} has been vouched for by ${voucher.name}!`,
+    circleId: nominee.circle_id,
+    sanitize: true,
+    channels: {
+      // TODO: figure out if these need to be conditionalized?
+      discord: true,
+      telegram: true,
+    },
+  });
+
+  // if there are enough nominations, go ahead and add the user to the circle
+  const nomCount = nominee.nominations_aggregate.aggregate?.count || 0;
+  if (nominee.vouches_required - 1 <= nomCount) {
+    return await convertNomineeToUser(nominee);
+  }
+
+  return nominee;
+}
+
+async function convertNomineeToUser(nominee: nominee) {
+  // Get the nominee into the user table
+  let userId = nominee.user_id;
+  if (!userId) {
+    const addedUser = await gql.insertUser(
+      nominee.address,
+      nominee.name,
+      nominee.circle_id
+    );
+    if (!addedUser) {
+      throw new InternalServerError('unable to add user');
+    }
+    userId = addedUser.id;
+  }
+
+  // Make sure they have a profile too
+  const { profiles } = await gql.getProfileAndMembership(nominee.address);
+  if (profiles.length == 0) {
+    const addedProfiles = await gql.insertProfiles([
+      {
+        address: nominee.address,
+      },
+    ]);
+    if (!addedProfiles) {
+      throw new InternalServerError('unable to add profile for new user');
+    }
+  }
+
+  // attach the user id to the nominee, and mark the nomination ended
+  const updatedNominee = await gql.updateNomineeUser(nominee.id, userId);
+  if (!updatedNominee) {
+    throw new InternalServerError('unable to update nominee userId');
+  }
+
+  // announce that they are vouched in
+  await sendSocialMessage({
+    message: `${nominee.name} has received enough vouches and is now in the circle!`,
+    circleId: nominee.circle_id,
+    sanitize: true,
+    channels: {
+      // TODO: figure out if these need to be conditionalized?
+      discord: true,
+      telegram: true,
+    },
+  });
+  return updatedNominee;
+}
+
+async function getNominee(nomineeId: number) {
+  const nom = await gql.getNominee(nomineeId);
+  if (!nom.nominees_by_pk) {
+    throw `nominee ${nomineeId} not found`;
+  }
+  return nom.nominees_by_pk;
+}
+
+type voucher = Awaited<ReturnType<typeof getUserFromProfileId>>;
+type nominee = Awaited<ReturnType<typeof getNominee>>;
```

Comment:
Types should be capitalized
```suggestion
type Voucher = Awaited<ReturnType<typeof getUserFromProfileId>>;
type Nominee = Awaited<ReturnType<typeof getNominee>>;
```